fluent-syntax/src/ast.rs

```diff
@@ -1,129 +1,134 @@
 }
 
 #[derive(Debug, PartialEq)]
-pub enum Expression {
-    StringExpression {
-        value: String,
-    },
-    NumberExpression {
-        value: Number,
-    },
-    MessageReference {
-        id: Identifier,
-    },
-    ExternalArgument {
-        id: Identifier,
-    },
-    SelectExpression {
-        expression: Option<Box<Expression>>,
-        variants: Vec<Variant>,
-    },
-    AttributeExpression {
-        id: Identifier,
-        name: Identifier,
-    },
-    VariantExpression {
-        id: Identifier,
-        key: VarKey,
-    },
-    CallExpression {
-        callee: Function,
-        args: Vec<Argument>,
-    },
+pub struct Pattern<'ast> {
+    pub elements: Vec<PatternElement<'ast>>,
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Attribute {
-    pub id: Identifier,
-    pub value: Pattern,
+pub enum PatternElement<'ast> {
+    TextElement(&'ast str),
+    Placeable(Expression<'ast>),
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Variant {
-    pub key: VarKey,
-    pub value: Pattern,
-    pub default: bool,
+pub struct Attribute<'ast> {
+    pub id: Identifier<'ast>,
+    pub value: Pattern<'ast>,
 }
 
 #[derive(Debug, PartialEq)]
-pub enum VarKey {
-    VariantName(VariantName),
-    Number(Number),
+pub struct Identifier<'ast> {
+    pub name: &'ast str,
 }
 
 #[derive(Debug, PartialEq)]
-pub enum Argument {
-    Expression(Expression),
-    NamedArgument { name: Identifier, val: ArgValue },
+pub struct Variant<'ast> {
+    pub key: VariantKey<'ast>,
+    pub value: Value<'ast>,
+    pub default: bool,
 }
 
 #[derive(Debug, PartialEq)]
-pub enum ArgValue {
-    Number(Number),
-    String(String),
-}
-
-#[derive(Debug, PartialEq, Eq, Hash)]
-pub struct Identifier {
-    pub name: String,
+pub enum VariantKey<'ast> {
+    Identifier { name: &'ast str },
+    NumberLiteral { value: &'ast str },
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Number {
-    pub value: String,
+pub enum Comment<'ast> {
+    Comment { content: Vec<&'ast str> },
+    GroupComment { content: Vec<&'ast str> },
+    ResourceComment { content: Vec<&'ast str> },
 }
 
 #[derive(Debug, PartialEq)]
-pub struct VariantName {
-    pub name: String,
+pub enum InlineExpression<'ast> {
+    StringLiteral {
+        raw: &'ast str,
+    },
+    NumberLiteral {
+        value: &'ast str,
+    },
+    VariableReference {
+        id: Identifier<'ast>,
+    },
+    CallExpression {
+        callee: Box<InlineExpression<'ast>>,
+        positional: Vec<InlineExpression<'ast>>,
+        named: Vec<NamedArgument<'ast>>,
+    },
+    AttributeExpression {
+        reference: Box<InlineExpression<'ast>>,
+        name: Identifier<'ast>,
+    },
+    VariantExpression {
+        reference: Box<InlineExpression<'ast>>,
+        key: VariantKey<'ast>,
+    },
+    MessageReference {
+        id: Identifier<'ast>,
+    },
+    TermReference {
+        id: Identifier<'ast>,
+    },
+    FunctionReference {
+        id: Identifier<'ast>,
+    },
```

Comment:
The `FunctionReference` isn't in the ebnf?

Reply:
The issue is in :

```ebnf
CallExpression ::= CalleeExpression blank? "(" blank? argument_list blank? ")"
CalleeExpression    ::= AttributeExpression
                      | FunctionReference
                      | TermReference
```

The `callee` field of `CallExpression` must be *something* and that something is at the moment made of two `InlineExpression` variants (`AttributeExpression` and `TermReference`) and a standalone `FunctionReference`.
What's more `FunctionReference` is not used anywhere else.

So, my options are:

*) Create a standalone node called `FunctionReference` and create a `CalleeExpression` enum that can take either `InlineExpression` or `FunctionReference` and continue guarding against invalid variants of `InlineExpression` being used in the `callee` field in the parser rather than ast.
*) Move both `AttributeExpression` and `TermReference` to be standalone Nodes that can be also used as variants of InlineExpression and then create `CalleeExpression` to reference them.
*) Do what I did here

The second option would look sth like this:

```rust
struct AttributeExpression {...}
struct TermReference {...}
struct FunctionReference {...}

enum InlineExpression {
  StringLiteral{ raw: &str },
  NumberLiteral{ value: &str },
  VariableReference{ id: Identifier },
  CallExpression {...},
  AttributeExpression(AttributeExpression),
  VariantExpression { ... },
  MessageReference { ... },
  TermReference(TermReference),
  Placeable { ... }
}

enum CalleeExpression {
  AttributeExpression(AttributeExpression),
  TermReference(TermReference),
  FunctionReference(FunctionReference), 
}
```

I think it's not worth the trouble. I'll comment in the AST.